Add app class to `urlMappingsReport` task args

matrei opened the pull request:
Closes #15286

matrei committed:
fix: add app class to `urlMappingsReport` task args

jdaugherty commented on grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/web/GrailsWebGradlePlugin.groovy:
```diff
@@ -46,19 +44,25 @@ class GrailsWebGradlePlugin extends GrailsGradlePlugin {
-                systemProperty(Environment.KEY, System.getProperty(Environment.KEY, Environment.DEVELOPMENT.getName()))
-                command = 'url-mappings-report'
-            }
+        project.tasks.register('urlMappingsReport', ApplicationContextCommandTask) { task ->
```
So why do we have to even register a custom command here?  Why isn't the UrlMappingsReportCommand being found by `configureApplicationCommands` in `GrailsGradlePlugin`?

matrei replied:
@jdaugherty Yes, you are right. I was not aware of this factory loading. I'll update the PR to utilize this functionality instead.

matrei replied:
So, the problem seems to be that `grails-web-url-mappings` which contains the `UrlMappingsReportCommand` class and also registers that class as a command in `META-INF/services/grails.factories` is currently not added to the build classpath in Grails applications.

The `SchemaExportCommand` from `grails-data-hibernate5` is picked up as we explicitly add it to the buildscript block.

matrei replied:
And it seems that the `META-INF/services/grails.factories` file in `grails-web-url-mappings` is moot as `GlobalGrailsClassInjectorTransformation` creates `META-INF/grails.factories` files dynamically.

jdaugherty replied:
Shouldn't the gradle web plugin then have it as an api dependency?

matrei replied:
> Shouldn't the gradle web plugin then have it as an api dependency?

I'm not sure you can add a dependency to only one of the published Gradle plugins. They all share the same dependencies.

It seems that the way these commands are handled by other modules is to add them to the build dependencies explicitly in the application. For example the `grails-data-hibernate5` and `grails-data-hibernate5-dbmigration` below:
```groovy
buildscript {
    repositories {
        mavenCentral()
        maven {
           url = 'https://repo.grails.org/grails/restricted'
        }
    }
    dependencies { // Not Published to Gradle Plugin Portal
        classpath "cloud.wondrify:asset-pipeline-gradle"
        classpath platform("org.apache.grails:grails-bom:$grailsVersion")
        classpath "org.apache.grails:grails-data-hibernate5"
        classpath "org.apache.grails:grails-data-hibernate5-dbmigration"
        classpath "org.apache.grails:grails-gradle-plugins"
    }
}
```

jdaugherty replied:
Ah, good point.  We can't do that since we have a central one.  Alright, let's leave it.

matrei committed:
fix: remove unused `grails.factories` file

Remove `META-INF/services/grails.factories` from `grails-web-url-mappings`.
The file explicitly registers `UrlMappingsReportCommand`, but this is
redundant: `GlobalGrailsClassInjectorTransformation` dynamically generates
`META-INF/grails.factories` and automatically registers all
`ApplicationCommand` implementations found in the source tree.

matrei committed:
chore: cleanup `ApplicationCommand` and `UrlMappingsReportCommand`

matrei committed:
fix: guard against multiple task registrations

If grails-web-urlmappings is on the build runtime classpath,
its service-loader initialization will pre-register the
`urlMappingsReport` task. This change adds a check to ensure
the task is only registered once.

matrei committed:
fix: handle missing configurations in buildClasspath setup

Use `getOrNull` and filter out null configurations to avoid
potential issues during build classpath generation.

matrei committed:
fix: compile error
